File: sdk/python/kflow/workflow.py
"""Workflow, TaskDef, StepBuilder, and sentinel constants."""
from __future__ import annotations

import inspect
from typing import Callable, Optional

# Sentinel constants — must match Go/Rust SDK values exactly.
Succeed = "__succeed__"
Fail    = "__fail__"

# Input/Output are plain dicts; all values must be JSON-serialisable.
Input  = dict
Output = dict


class _RetryPolicy:
    def __init__(self, max_attempts: int, backoff_seconds: int = 0) -> None:
        self.max_attempts    = max_attempts
        self.backoff_seconds = backoff_seconds


class TaskDef:
    def __init__(self, name: str) -> None:
        self._name           = name
        self._handler        = None   # sync or async callable
        self._choice_fn      = None
        self._is_choice      = False
        self._is_wait        = False
        self._wait_seconds   = 0
        self._service_target: Optional[str] = None
        self._retry: Optional[_RetryPolicy]  = None
        self._catch: Optional[str]           = None

    def invoke_service(self, service_name: str) -> "TaskDef":
        self._service_target = service_name
        return self

    def retry(self, max_attempts: int, backoff_seconds: int = 0) -> "TaskDef":
        self._retry = _RetryPolicy(max_attempts, backoff_seconds)
        return self

    def catch(self, state_name: str) -> "TaskDef":
        self._catch = state_name
        return self


class StepBuilder:
    def __init__(self, name: str) -> None:
        self.name              = name
        self._next: Optional[str]          = None
        self._catch: Optional[str]         = None
        self._retry: Optional[_RetryPolicy] = None
        self._is_end           = False

    def next(self, state_name: str) -> "StepBuilder":
        self._next = state_name
        return self

    def catch(self, state_name: str) -> "StepBuilder":
        self._catch = state_name
        return self

    def retry(self, max_attempts: int, backoff_seconds: int = 0) -> "StepBuilder":
        self._retry = _RetryPolicy(max_attempts, backoff_seconds)
        return self

    def end(self) -> "StepBuilder":
        self._next   = Succeed
        self._is_end = True
        return self


def step(name: str) -> StepBuilder:
    """Convenience constructor for StepBuilder."""
    return StepBuilder(name)


class Workflow:
    def __init__(self, name: str) -> None:
        self._name  = name
        self._image = ""
        self._tasks: dict[str, TaskDef] = {}
        self._names: list[str] = []   # insertion-order duplicate tracking
        self._steps: list[StepBuilder] = []

    @property
    def name(self) -> str:
        return self._name

    def with_image(self, image: str) -> "Workflow":
        """Set the container image for K8s Job execution. Empty = in-process."""
        self._image = image
        return self

    def task(self, name: str) -> Callable:
        """Decorator — registers a function as a Task state."""
        def decorator(fn: Callable) -> Callable:
            td = TaskDef(name)
            if inspect.iscoroutinefunction(fn):
                td._handler = fn
            else:
                td._handler = fn
            self._tasks[name] = td
            self._names.append(name)
            return fn
        return decorator

    def choice(self, name: str) -> Callable:
        """Decorator — registers a function as a Choice state."""
        def decorator(fn: Callable) -> Callable:
            td = TaskDef(name)
            td._is_choice = True
            td._choice_fn = fn
            self._tasks[name] = td
            self._names.append(name)
            return fn
        return decorator

    def wait(self, name: str, seconds: int) -> TaskDef:
        td = TaskDef(name)
        td._is_wait      = True
        td._wait_seconds = seconds
        self._tasks[name] = td
        self._names.append(name)
        return td

    def flow(self, *steps: StepBuilder) -> None:
        self._steps = list(steps)

    def validate(self) -> None:
        """Raises ValueError if any invariant is violated."""
        # 1. flow() must have been called with at least one step.
        if not self._steps:
            raise ValueError("flow() has not been called or contains no steps")

        # 2. All state names must be globally unique.
        seen: set[str] = set()
        for n in self._names:
            if n in seen:
                raise ValueError(f"duplicate state name: {n!r}")
            seen.add(n)

        # 3. Each task has exactly one of: handler or invoke_service.
        for name, td in self._tasks.items():
            if td._is_choice or td._is_wait:
                continue
            has_handler = td._handler is not None
            has_service = td._service_target is not None
            if has_handler and has_service:
                raise ValueError(
                    f"state {name!r}: cannot have both a handler and invoke_service"
                )
            if not has_handler and not has_service:
                raise ValueError(
                    f"state {name!r}: must have either a handler or invoke_service"
                )

        # 4. Every next/catch target must resolve to a registered state, Succeed, or Fail.
        valid_targets = set(self._tasks.keys()) | {Succeed, Fail}
        for sb in self._steps:
            if sb._next and sb._next not in valid_targets:
                raise ValueError(
                    f"step {sb.name!r}: next target {sb._next!r} is not a registered state"
                )
            if sb._catch and sb._catch not in valid_targets:
                raise ValueError(
                    f"step {sb.name!r}: catch target {sb._catch!r} is not a registered state"
                )
            # task-level catch
            td = self._tasks.get(sb.name)
            if td and td._catch and td._catch not in valid_targets:
                raise ValueError(
                    f"state {sb.name!r}: catch target {td._catch!r} is not a registered state"
                )

File: sdk/rust/src/workflow.rs
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use crate::{Error, Input, Output};

/// Terminal success sentinel — must match Go/Python SDK values exactly.
pub const SUCCEED: &str = "__succeed__";

/// Terminal failure sentinel — must match Go/Python SDK values exactly.
pub const FAIL: &str = "__fail__";

pub type HandlerFn = Box<
    dyn Fn(Input) -> Pin<Box<dyn Future<Output = Result<Output, Error>> + Send>>
        + Send
        + Sync,
>;

pub type ChoiceFn = Box<
    dyn Fn(Input) -> Pin<Box<dyn Future<Output = Result<String, Error>> + Send>>
        + Send
        + Sync,
>;

#[derive(Debug, Clone, Default)]
pub struct RetryPolicy {
    pub max_attempts:    u32,
    pub backoff_seconds: u64,
}

pub(crate) enum StateKind {
    Task,
    Choice,
    Wait(Duration),
}

pub struct TaskDef {
    pub(crate) name:           String,
    pub(crate) handler:        Option<HandlerFn>,
    pub(crate) choice_handler: Option<ChoiceFn>,
    pub(crate) service_target: Option<String>,
    pub(crate) retry:          Option<RetryPolicy>,
    pub(crate) catch:          Option<String>,
    pub(crate) kind:           StateKind,
}

impl TaskDef {
    fn new_task(name: String, handler: Option<HandlerFn>) -> Self {
        TaskDef {
            name,
            handler,
            choice_handler: None,
            service_target: None,
            retry:          None,
            catch:          None,
            kind:           StateKind::Task,
        }
    }

    fn new_choice(name: String, handler: ChoiceFn) -> Self {
        TaskDef {
            name,
            handler:        None,
            choice_handler: Some(handler),
            service_target: None,
            retry:          None,
            catch:          None,
            kind:           StateKind::Choice,
        }
    }

    fn new_wait(name: String, dur: Duration) -> Self {
        TaskDef {
            name,
            handler:        None,
            choice_handler: None,
            service_target: None,
            retry:          None,
            catch:          None,
            kind:           StateKind::Wait(dur),
        }
    }

    pub fn invoke_service(&mut self, service_name: impl Into<String>) -> &mut Self {
        self.service_target = Some(service_name.into());
        self
    }

    pub fn retry(&mut self, policy: RetryPolicy) -> &mut Self {
        self.retry = Some(policy);
        self
    }

    pub fn catch(&mut self, state_name: impl Into<String>) -> &mut Self {
        self.catch = Some(state_name.into());
        self
    }

    pub fn is_choice(&self) -> bool { matches!(self.kind, StateKind::Choice) }
    pub fn is_wait(&self) -> bool   { matches!(self.kind, StateKind::Wait(_)) }

    pub fn wait_duration(&self) -> Option<Duration> {
        if let StateKind::Wait(d) = &self.kind { Some(*d) } else { None }
    }
}

/// Returns a new StepBuilder for the named state.
pub fn step(name: impl Into<String>) -> StepBuilder {
    StepBuilder::new(name.into())
}

#[derive(Clone)]
pub struct StepBuilder {
    pub(crate) name:   String,
    pub(crate) next:   Option<String>,
    pub(crate) catch:  Option<String>,
    pub(crate) retry:  Option<RetryPolicy>,
    pub(crate) is_end: bool,
}

impl StepBuilder {
    pub fn new(name: String) -> Self {
        StepBuilder { name, next: None, catch: None, retry: None, is_end: false }
    }

    pub fn next(mut self, state_name: impl Into<String>) -> Self {
        self.next = Some(state_name.into());
        self
    }

    pub fn catch(mut self, state_name: impl Into<String>) -> Self {
        self.catch = Some(state_name.into());
        self
    }

    pub fn retry(mut self, policy: RetryPolicy) -> Self {
        self.retry = Some(policy);
        self
    }

    /// Equivalent to .next(SUCCEED).
    pub fn end(mut self) -> Self {
        self.next   = Some(SUCCEED.to_string());
        self.is_end = true;
        self
    }
}

pub struct Workflow {
    pub(crate) name:  String,
    pub(crate) image: String,
    pub(crate) tasks: HashMap<String, TaskDef>,
    pub(crate) names: Vec<String>, // insertion-order for duplicate detection
    pub(crate) steps: Vec<StepBuilder>,
}

impl Workflow {
    pub fn new(name: impl Into<String>) -> Self {
        Workflow {
            name:  name.into(),
            image: String::new(),
            tasks: HashMap::new(),
            names: Vec::new(),
            steps: Vec::new(),
        }
    }

    /// Set the container image for K8s Job execution. Empty = in-process.
    pub fn with_image(mut self, image: impl Into<String>) -> Self {
        self.image = image.into();
        self
    }

    pub fn task(&mut self, name: impl Into<String>, handler: Option<HandlerFn>) -> &mut TaskDef {
        let n = name.into();
        let td = TaskDef::new_task(n.clone(), handler);
        self.names.push(n.clone());
        self.tasks.insert(n.clone(), td);
        self.tasks.get_mut(&n).unwrap()
    }

    pub fn choice(&mut self, name: impl Into<String>, handler: ChoiceFn) -> &mut TaskDef {
        let n = name.into();
        let td = TaskDef::new_choice(n.clone(), handler);
        self.names.push(n.clone());
        self.tasks.insert(n.clone(), td);
        self.tasks.get_mut(&n).unwrap()
    }

    pub fn wait(&mut self, name: impl Into<String>, duration: Duration) -> &mut TaskDef {
        let n = name.into();
        let td = TaskDef::new_wait(n.clone(), duration);
        self.names.push(n.clone());
        self.tasks.insert(n.clone(), td);
        self.tasks.get_mut(&n).unwrap()
    }

    pub fn flow(&mut self, steps: Vec<StepBuilder>) -> &mut Self {
        self.steps = steps;
        self
    }

    pub fn validate(&self) -> Result<(), String> {
        // 1. flow() must have at least one step.
        if self.steps.is_empty() {
            return Err("flow() has not been called or contains no steps".to_string());
        }

        // 2. Globally unique state names.
        let mut seen = std::collections::HashSet::new();
        for n in &self.names {
            if !seen.insert(n) {
                return Err(format!("duplicate state name: {n:?}"));
            }
        }

        // 3. Each task has exactly one of: handler or invoke_service.
        for (name, td) in &self.tasks {
            if td.is_choice() || td.is_wait() { continue; }
            let has_handler  = td.handler.is_some();
            let has_service  = td.service_target.is_some();
            if has_handler && has_service {
                return Err(format!(
                    "state {name:?}: cannot have both a handler and invoke_service"
                ));
            }
            if !has_handler && !has_service {
                return Err(format!(
                    "state {name:?}: must have either a handler or invoke_service"
                ));
            }
        }

        // 4. Every next/catch target resolves to a registered state, SUCCEED, or FAIL.
        let valid: std::collections::HashSet<&str> = self
            .tasks
            .keys()
            .map(|s| s.as_str())
            .chain([SUCCEED, FAIL])
            .collect();

        for sb in &self.steps {
            if let Some(n) = &sb.next {
                if !valid.contains(n.as_str()) {
                    return Err(format!(
                        "step {:?}: next target {n:?} is not a registered state",
                        sb.name
                    ));
                }
            }
            if let Some(c) = &sb.catch {
                if !valid.contains(c.as_str()) {
                    return Err(format!(
                        "step {:?}: catch target {c:?} is not a registered state",
                        sb.name
                    ));
                }
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn handler() -> Option<HandlerFn> {
        Some(Box::new(|input| Box::pin(async move { Ok(input) })))
    }

    fn choice_fn() -> ChoiceFn {
        Box::new(|_| Box::pin(async move { Ok(SUCCEED.to_string()) }))
    }

    #[test]
    fn test_sentinels() {
        assert_eq!(SUCCEED, "__succeed__");
        assert_eq!(FAIL,    "__fail__");
    }

    #[test]
    fn test_with_image_sets_field() {
        let wf = Workflow::new("test").with_image("my-image:latest");
        assert_eq!(wf.image, "my-image:latest");
    }

    #[test]
    fn test_image_default_empty() {
        let wf = Workflow::new("test");
        assert!(wf.image.is_empty());
    }

    #[test]
    fn test_validate_ok() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler());
        wf.flow(vec![step("A").end()]);
        assert!(wf.validate().is_ok());
    }

    #[test]
    fn test_validate_no_flow() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler());
        assert!(wf.validate().is_err());
    }

    #[test]
    fn test_validate_duplicate_state() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler());
        wf.task("A", handler()); // duplicate
        wf.flow(vec![step("A").end()]);
        assert!(wf.validate().unwrap_err().contains("duplicate"));
    }

    #[test]
    fn test_validate_handler_and_service() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler()).invoke_service("svc");
        wf.flow(vec![step("A").end()]);
        assert!(wf.validate().unwrap_err().contains("both"));
    }

    #[test]
    fn test_validate_no_handler_no_service() {
        let mut wf = Workflow::new("test");
        wf.task("A", None); // no handler, no service
        wf.flow(vec![step("A").end()]);
        assert!(wf.validate().unwrap_err().contains("must have"));
    }

    #[test]
    fn test_validate_unknown_next_target() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler());
        wf.flow(vec![step("A").next("NonExistent")]);
        assert!(wf.validate().unwrap_err().contains("not a registered state"));
    }

    #[test]
    fn test_validate_unknown_catch_target() {
        let mut wf = Workflow::new("test");
        wf.task("A", handler());
        wf.flow(vec![step("A").catch("Ghost").end()]);
        assert!(wf.validate().unwrap_err().contains("not a registered state"));
    }

    #[test]
    fn test_validate_choice_ok() {
        let mut wf = Workflow::new("test");
        wf.choice("Route", choice_fn());
        wf.task("Next", handler());
        wf.flow(vec![step("Route").next("Next"), step("Next").end()]);
        assert!(wf.validate().is_ok());
    }
}
